pkg/k8s/statedb.go

```diff
@@ -0,0 +1,313 @@
+// Copyright Authors of Cilium
+
+package k8s
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+	"github.com/cilium/stream"
+
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// RegisterReflector registers a Kubernetes to StateDB table reflector.
+//
+// Intended to be used with [cell.Invoke] and the module's job group.
+// See [ExampleRegisterReflector] for example usage.
+func RegisterReflector[Obj any](jobGroup job.Group, db *statedb.DB, targetTable statedb.RWTable[Obj], cfg ReflectorConfig[Obj]) {
+	// Register initializer that marks when the table has been initially populated,
+	// e.g. the initial "List" has concluded.
+	r := &k8sReflector[Obj]{
+		ReflectorConfig: cfg.withDefaults(),
+		db:              db,
+		table:           targetTable,
+	}
+	wtxn := db.WriteTxn(targetTable)
+	r.initDone = targetTable.RegisterInitializer(wtxn, "k8s-reflector")
+	wtxn.Commit()
+
+	jobGroup.Add(job.OneShot(
+		fmt.Sprintf("k8s-reflector-[%T]", *new(Obj)),
+		r.run))
+}
+
+type ReflectorConfig[Obj any] struct {
+	// Maximum number of objects to commit in one transaction. Uses default if left zero.
+	BufferSize int
+
+	// The amount of time to wait for the buffer to fill. Uses default if left zero.
+	BufferWaitTime time.Duration
+
+	// The ListerWatcher to use to retrieve the objects.
+	//
+	// Use [utils.ListerWatcherFromTyped] to create one from the Clientset, e.g.
+	//
+	//   var cs client.Clientset
+	//   utils.ListerWatcherFromTyped(cs.CoreV1().Nodes())
+	//
+	ListerWatcher cache.ListerWatcher
+
+	// Optional function to transform the objects given by the ListerWatcher. This can
+	// be used to convert into an internal model on the fly to save space and add additional
+	// fields or to for example implement TableRow/TableHeader for a cilium-dbg statedb command.
+	Transform TransformFunc[Obj]
+
+	// Optional function to query all objects. Used when replacing the objects on resync.
+	// This can be used to "namespace" the objects managed by this reflector, e.g. on
+	// source.Source etc.
+	QueryAll QueryAllFunc[Obj]
+}
+
+// TransformFunc is an optional function to give to the Kubernetes reflector
+// to transform the object returned by the ListerWatcher to the desired
+// target object. If the function returns false the object is silently
+// skipped.
+type TransformFunc[Obj any] func(any) (obj Obj, ok bool)
+
+// QueryAllFunc is an optional function to give to the Kubernetes reflector
+// to query all objects in the table that are managed by the reflector.
+// It is used to delete all objects when the underlying cache.Reflector needs
+// to Replace() all items for a resync.
+type QueryAllFunc[Obj any] func(statedb.ReadTxn, statedb.Table[Obj]) statedb.Iterator[Obj]
+
+const (
+	// DefaultBufferSize is the maximum number of objects to commit to the table in one write transaction.
+	DefaultBufferSize = 256
```

Comment:
TODO raise the defaults to get something like 50k/s. This is 5120 objects per second.